scheduler/src/cook/mesos/api.clj

```diff
@@ -213,31 +221,38 @@
    ;; Make sure the user name is valid. It must begin with a lower case character, end with
    ;; a lower case character or a digit, and has length between 2 to (62 + 2).
    :user UserName
-   (s/optional-key :application) Application})
+   (s/optional-key :application) Application
+   (s/optional-key :expected-runtime) PosInt})
+
+(def Job
+  "Full schema for a job"
+  (s/constrained JobMap valid-runtimes?))
+
+(def JobRequestMap
+  "Schema for the fields of a job request"
+  (-> JobMap
+      (dissoc :name)
+      (dissoc :priority)
+      (dissoc :max-runtime)
+      (dissoc :user)
+      (merge {(s/optional-key :name) JobName
+              (s/optional-key :priority) JobPriority
+              (s/optional-key :uris) [UriRequest]
+              (s/optional-key :env) {s/Keyword s/Str}
+              (s/optional-key :labels) {s/Keyword s/Str}
+              (s/optional-key :max-runtime) PosInt
+              :cpus PosNum
+              :mem PosNum})))
 
 (def JobRequest
   "Schema for the part of a request that launches a single job."
-  (-> Job
-    ;; make max-runtime optional.
-    ;; It is *not* optional internally but don't want to force users to set it
-    (dissoc :name)
-    (dissoc :priority)
-    (dissoc :max-runtime)
-    (dissoc :user)
-    (merge {(s/optional-key :name) JobName
-            (s/optional-key :priority) JobPriority
-            (s/optional-key :uris) [UriRequest]
-            (s/optional-key :env) {s/Keyword s/Str}
-            (s/optional-key :labels) {s/Keyword s/Str}
-            (s/optional-key :max-runtime) PosInt
-            :cpus PosNum
-            :mem PosNum})))
+  (s/constrained JobRequestMap valid-runtimes?))
```

Comment:
Because in the `JobResponse` schema def below, we are `dissoc`ing and `merge`ing with the `JobRequestMap`, so the additional `valid-runtimes?` constraint shouldn't be present there.